### PR TITLE
Make fewer types generic over QueryContext

### PR DESCRIPTION
**Pull request by Julian-Wollersberger**
While trying to refactor `rustc_query_system::query::QueryContext` to make it dyn-safe, I noticed some smaller things:
* QueryConfig doesn't need to be generic over QueryContext
* ~~The `kind` field on QueryJobId is unused~~
* Some unnecessary where clauses
* Many types in `job.rs` where generic over `QueryContext` but only needed `QueryContext::Query`.
  If handle_cycle_error() could be refactored to not take `error: CycleError<CTX::Query>`, all those bounds could be removed as well.

Changing `find_cycle_in_stack()` in job.rs to not take a `tcx` argument is the only functional change here. Everything else is just updating type signatures. (aka compile-error driven development ^^)

~~Currently there is a weird bug where memory usage suddenly skyrockets when running UI tests. I'll investigate that tomorrow.
A perf run probably won't make sense before that is fixed.~~

EDIT: `kind` actually is used by `Eq`, and re-adding it fixed the memory issue.

### CODE REVIEW

**Comment by varkor** [comment]
I'm not very familiar with the query system, so someone like r? @oli-obk would be a better reviewer.

**Comment by mati865** [comment]
Slight regression across the board.

File: compiler/rustc_query_system/src/query/job.rs [hunk 1]
```diff
@@ -29,73 +30,77 @@ pub struct QueryInfo<Q> {
...
+pub struct QueryJobId<D> {
     /// Which job within a shard is this
     pub job: QueryShardJobId,
 
     /// In which shard is this job
     pub shard: u16,
+
+    /// What kind of query this job is.
+    //FIXME: This seems to be unused, but removing it changes query cycle errors.
+    pub kind: D,
```
Comment (cjgillot) [inline compiler/rustc_query_system/src/query/job.rs:50]: What if its used by the `Eq` impl?

Comment (Julian-Wollersberger) [reply compiler/rustc_query_system/src/query/job.rs:50]: Ohhhhhh, that must be it! Thanks!
I just didn't understand what was going on.

**Comment by Julian-Wollersberger** [comment]
This PR has given me headaches, but now it should be ready.

First, I undid the removal of the `kind` field of `QueryJobId`. This one was the cause of the memory bug, 
~~but I still don't understand how that happened. The field wasn't read anywhere.~~
EDIT: As @cjgillot mentioned, it's probably the `#[derive(Eq)]`.

Measuring where the perf regression came from was also tricky, because my two identical checkouts of rustc turned out to not be identical. Sight. Passing `Option<QueryJobId>` by value turned out to be the culprit.

**Comment by cjgillot** [comment]
This crate needs some convention for the type parameters. The  amount of those is getting out of hand.

**Comment by Julian-Wollersberger** [comment]
I feel you, since I've been doing type juggling for hours on end. 
Let's see if I can make `kind: Box<dyn DepKind>` in QueryJobId.

**Comment by Julian-Wollersberger** [comment]
Performance is neutral.

I looked into `dyn DepKind`, which was also suggested by @oli-obk on Zulip, but it's not easy since `DepKind` implements `Clone + Eq + Ord + Hash` which are all not dyn safe.
So it's probably better to do that in a separate PR.

With that, can you review it @oli-obk?
It's probably best to review the total changes, since three commits mostly touch the same lines. It's mostly type signature changes, and I changed the arguments to `find_cycle_in_stack()` and `pick_query()` in `job.rs`.

File: compiler/rustc_query_system/src/query/plumbing.rs [hunk 2]
```diff
@@ -467,17 +446,14 @@ where
...
+fn load_from_disk_and_cache_in_memory<CTX: QueryContext, K, V>(
     tcx: CTX,
     key: K,
     prev_dep_node_index: SerializedDepNodeIndex,
     dep_node_index: DepNodeIndex,
     dep_node: &DepNode<CTX::DepKind>,
     query: &QueryVtable<CTX, K, V>,
-) -> V
-where
-    CTX: QueryContext,
```
Comment (oli-obk) [inline compiler/rustc_query_system/src/query/plumbing.rs:479]: All these changes moving bounds from where bounds to generic arg declarations are not necessary for this PR. Please revert them. If I remember correctly this is all on purpose, let's not change it in a drive-by manner.

**Comment by oli-obk** [comment]
Except for the stylistic nit, this lgtm

**Comment by cjgillot** [comment]
Could you remove the two `QueryJobId` commits? The back-and-forth changes the type parameters needlessly.

File: compiler/rustc_query_system/src/query/plumbing.rs (hunk 2 above)
Comment (Julian-Wollersberger) [reply compiler/rustc_query_system/src/query/plumbing.rs:479]: Should I also change it when there was no where clause but now the generic args are a lot longer?
Like
```
impl<D: Copy + Clone + Eq + Hash, Q: Clone> QueryJob<D, Q> {
```
to
```
impl<D, Q> QueryJob<D, Q>
where
    D: Copy + Clone + Eq + Hash,
    Q: Clone,
```

**Comment by Julian-Wollersberger** [comment]
> Could you remove the two `QueryJobId` commits? The back-and-forth changes the type parameters needlessly.

The type is changed from `CXT: QueryContext` to `D`.
If it isn't to annoying to you, I wouldn't do it because squashing the commits would be a pain. Although CLion has this neat "Interactively rebase from here" feature, I'd still have to resolve merge conflicts in 7 files multiplied by 3 commits.

**Comment by oli-obk** [comment]
> because squashing the commits would be a pain

Why is squashing a problem? Or do you mean, move the commits around, then squash? Just straight sqashing sequential commits never causes any merge conflicts

**Comment by Julian-Wollersberger** [comment]
Yes, I meant moving commits around. I was concerned because otherwise it would squash to many commits.
So I guess it's better to squash to much than to little? 

Anyway, I squashed everything but the first commit. Sorry for the back and forth.

**Comment by oli-obk** [comment]
@bors r+

thanks!

**Comment by Julian-Wollersberger** [comment]
This PR could be rolled up, right? Because it's perf neutral

**Comment by oli-obk** [comment]
@bors rollup=iffy

sure, but it also touches a lot of stuff, so it conflicts a lot
